dora/src/language/generator.rs

```diff
@@ -871,40 +879,60 @@ impl<'a> AstBytecodeGen<'a> {
                 self.ensure_register(dest, BytecodeType::from_ty(self.sa, ty))
             };
 
-            let else_lbl = self.gen.create_label();
+            let mut next_lbl;
             let end_lbl = self.gen.create_label();
 
-            let cond_reg = self.visit_expr(&expr.cond, DataDest::Alloc);
-            self.gen.emit_jump_if_false(cond_reg, else_lbl);
-            self.free_if_temp(cond_reg);
-
-            self.visit_expr(&expr.then_block, DataDest::Reg(dest));
+            let cond_head_reg = self.visit_stmt_let(&expr.cond_head).unwrap();
 
-            if !expr_always_returns(&expr.then_block) {
-                self.gen.emit_jump(end_lbl);
+            for branch in &expr.branches {
+                next_lbl = self.gen.create_label();
+                self.visit_if_branch(cond_head_reg, branch, next_lbl, end_lbl, dest);
             }
 
-            self.gen.bind_label(else_lbl);
             self.visit_expr(else_block, DataDest::Reg(dest));
             self.gen.bind_label(end_lbl);
 
             dest
         } else {
-            // Without else-branch there can't be return value
+            // Without else-branch there can't be return value --- not true anymore, if cond_head is an enum, branches can be exhaustive
```

Comment:
The comment is probably a bit premature, as support for enums hasn't been implemented yet.

Reply:
Reworded.